backend/de.metas.contracts/src/main/java/de/metas/contracts/pricing/trade_margin/CustomerTradeMarginService.java

```diff
@@ -22,97 +22,211 @@
-						continue;
-					}
-
-					final int commissionPointsPrecision = HierarchyContract.cast(commissionContract).getPointsPrecision();
-
-					final CommissionPoints tradedCommissionPoints = salesRepCommissionPoints
-							.computePercentageOf(
-									Percent.of(customerTradeMarginSettings.getMarginPercent()),
-									commissionPointsPrecision);
-
-					logger.debug("put commission-share with tradedCommissionPoints={} to the overall result", tradedCommissionPoints);
-					result.put(commissionShare, tradedCommissionPoints);
-				}
-			}
-
-			return result.build();
+			loggable.addLog("createSalesRepPricingResult - No priceListId set for bpartner={}" + request.getSalesRepId());
+			return Optional.empty();
 		}
+
+		final OrgId salesRepOrgId = OrgId.ofRepoId(salesRepBillToLocation.getAD_Org_ID());
+
+		final IEditablePricingContext salesRepPricingCtx =
+				pricingBL.createInitialContext(salesRepOrgId,
+											   request.getProductId(),
+											   request.getSalesRepId(),
+											   request.getQty(),
+											   request.getSoTrx())
+						.setPricingSystemId(pricingSystemId)
+						.setCountryId(countryId)
+						.setPriceListId(priceListId)
+						.setConvertPriceToContextUOM(false);
+
+		final IPricingResult salesRepPriceResult = pricingBL.calculatePrice(salesRepPricingCtx);
+
+		if (!salesRepPriceResult.isCalculated())
+		{
+			loggable.addLog("createSalesRepPricingResult - Price not calculated for bpartner={}" + request.getSalesRepId());
+			return Optional.empty();
+		}
+
+		updatePricingResultToMatchUOM(salesRepPriceResult, request.getQty().getUomId());
+
+		final Money salesRepNetUnitPriceWithoutTax = deductTaxes(salesRepOrgId,
+													   salesRepBillToLocation,
+													   request,
+													   salesRepPriceResult);
+
+		final Money salesRepNetUnitPrice = convertToCustomerCurrency(salesRepOrgId, request, salesRepNetUnitPriceWithoutTax);
+
+		return Optional.of(ProductPrice.builder()
+								   .money(salesRepNetUnitPrice)
+								   .productId(salesRepPriceResult.getProductId())
+								   .uomId(salesRepPriceResult.getPriceUomId())
+								   .build());
+	}
+
+	@NonNull
+	private Money deductTaxes(
+			@NonNull final OrgId salesRepOrgId,
+			@NonNull final I_C_BPartner_Location salesRepBillToLocation,
+			@NonNull final ComputeSalesRepPriceRequest request,
+			@NonNull final IPricingResult salesRepPricingResult)
+	{
+		final BPartnerLocationId salesRepBillToLocationId = BPartnerLocationId.ofRepoId(salesRepBillToLocation.getC_BPartner_ID(),
+																						salesRepBillToLocation.getC_BPartner_Location_ID());
+
+		final BPartnerLocationAndCaptureId salesRepBillToLocationAndCapture = BPartnerLocationAndCaptureId.of(salesRepBillToLocationId,
+																											  LocationId.ofRepoId(salesRepBillToLocation.getC_Location_ID()));
+
+		final TaxId taxId = taxBL.getTaxNotNull(
```

Comment:
Instead of querying the tax, better make sure that we have the tax from the underlying sales-order or commission-trigger.
(or let me know why querying makes more sense :-) )

Reply:
well, we don't have any sales-order / commission-trigger for our sales rep... we are just calculating which price he would get if it were for him to buy that product

Reply:
right